[doc] Add sphinx-autobuild and documentation for make local (#47275)
ray-project/ray: doc/source/ray-contribute/docs.md

Don't use `-U` in this step. You don't want to upgrade dependencies because `requirements-doc.txt` pins exact versions you need to build the docs.

### Build documentation

> Before building, you can clean your environment first by running:

Reviewer: ```suggestion
Before building, clean your environment first by running:
```

```shell
make clean
```

> Then, there are 2 options to build documentation locally that you can choose from:

Reviewer: ```suggestion
Choose from the following 2 options to build documentation locally:

- Incremental build
- Full build

```

Reviewer: Could you add some guidance on when you can't use the incremental build?

#### 1. Incremental build with global cache

> In this approach, we only build the changes you've made in your branch compared to your last pull from upstream master. The rest is cached with pre-built doc pages from your last upstream pull. (for every new commit pushed to Ray, we build all the documentation pages from that commit and store them on S3 as cache)

Reviewer: ```suggestion

Use incremental builds if you need to make frequent uncomplicated changes.
In this approach, the incremental build only builds the changes you made in your branch compared to your last pull from upstream master. The incremental build caches the rest cache with pre-built doc pages from your last upstream pull. For every new commit pushed to Ray, CI builds all the documentation pages from that commit and store them on S3 as cache.
```

> To do this, we first trace your commit tree to find the latest commit that we already cached on S3. 

Reviewer: ```suggestion
The incremental build first traces your commit tree to find the latest commit that CI already cached on S3. 
```

> Once the commit is found, we fetch its corresponding cache from S3 and extract into the `doc/` directory. Simultaneously, we track all the files that have changed from that commit to current `HEAD`, including any unstaged changes.

Reviewer: ```suggestion
Once the build finds the commit, it fetches the corresponding cache from S3 and extracts it into the `doc/` directory. Simultaneously, CI tracks all the files that have changed from that commit to current `HEAD`, including any unstaged changes.
```

Reviewer: Please check for accuracy. Avoid the use of "we" to reduce confusion. Please consider using Vale, which will catch these style issues at go/vale.

> Sphinx then rebuilds only the pages that depend on your changes, leaving the rest untouched from the cache.

Reviewer: ```suggestion
Sphinx then rebuilds only the pages that your changes affect, leaving the rest untouched from the cache.
```

> We highly recommend this option if you need to make frequent uncomplicated changes.

Reviewer: ```suggestion
```

To use this option, you can run:
```shell

> make local

Reviewer: Move this to the top of the section because this is what many folks will look for.

```

#### 2. Full build from scratch

> With this option, Sphinx would rebuild all files in `doc/` directory, ignoring all cache and saved environment.

Reviewer: ```suggestion
In the full build option, Sphinx rebuilds all files in `doc/` directory, ignoring all cache and saved environment.
```

> Because of this, you would get a really clean build but it would be much slower at the same time.

Reviewer: ```suggestion
Because of this behavior, you get a really clean build but it's much slower.
```
